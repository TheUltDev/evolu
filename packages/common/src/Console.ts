/**
 * üìù Cross-Platform Console
 *
 * Console abstraction for Chrome 123+, Firefox 125+, Safari 18.1+, Node.js
 * 22.x+, and React Native 0.75+. Includes methods guaranteed to be available in
 * these environments and expected to remain compatible in future versions.
 * Output formatting may vary (e.g., interactive UI in browsers vs. text in
 * Node.js/React Native), but functionality is consistent across platforms.
 *
 * **Warning**: If you encounter platform-specific issues or missing methods,
 * please contribute a PR with details about the environment and behavior.
 *
 * @module
 */

/**
 * Cross-platform Console interface for Chrome 123+, Firefox 125+, Safari 18.1+,
 * Node.js 22.x+, and React Native 0.75+
 */
export interface Console {
  /** Controls whether console methods produce output (default: true) */
  enabled?: boolean;

  /** Outputs a message to the console */
  log: (...args: Array<any>) => void;

  /** Outputs an informational message (often same as log) */
  info: (...args: Array<any>) => void;

  /** Outputs a warning message */
  warn: (...args: Array<any>) => void;

  /** Outputs an error message */
  error: (...args: Array<any>) => void;

  /** Outputs a debug message */
  debug: (...args: Array<any>) => void;

  /** Starts a timer with an optional label */
  time: (label?: string) => void;

  /** Ends a timer and logs the elapsed time */
  timeEnd: (label?: string) => void;

  /** Displays an object's properties in a detailed format */
  dir: (object: any, options?: any) => void;

  /** Displays tabular data as a table */
  table: (tabularData: any, properties?: Array<string>) => void;

  /** Logs the number of times this has been called with the given label */
  count: (label?: string) => void;

  /** Resets the counter for the given label */
  countReset: (label?: string) => void;

  /** Writes a message if the value is falsy, otherwise does nothing */
  assert: (value: any, message?: string, ...optionalParams: Array<any>) => void;

  /** Logs the elapsed time for a timer without ending it */
  timeLog: (label?: string, ...data: Array<any>) => void;

  /** Prints a stack trace with an optional message */
  trace: (message?: any, ...optionalParams: Array<any>) => void;
}

/** Dependency interface for injecting a Console instance. */
export interface ConsoleDep {
  readonly console: Console;
}

export interface ConsoleConfig {
  /**
   * Enable or disable console logging. When true, logs are output to the
   * {@link Console}; when false, logging is disabled for all methods except
   * `error`, which always outputs to ensure critical issues are not missed.
   * Default is false.
   */
  readonly enableLogging?: boolean;
}

/** Creates a console instance using the global console. */
export const createConsole = (config: ConsoleConfig = {}): Console => {
  const instance: Console = {
    enabled: config.enableLogging ?? false,

    log: (...args) => {
      // eslint-disable-next-line no-console, @typescript-eslint/no-unsafe-argument
      if (instance.enabled) console.log(...args);
    },
    info: (...args) => {
      // eslint-disable-next-line no-console, @typescript-eslint/no-unsafe-argument
      if (instance.enabled) console.info(...args);
    },
    warn: (...args) => {
      // eslint-disable-next-line no-console, @typescript-eslint/no-unsafe-argument
      if (instance.enabled) console.warn(...args);
    },
    error: (...args) => {
      // Always log errors, even if disabled
      // eslint-disable-next-line no-console, @typescript-eslint/no-unsafe-argument
      console.error(...args);
    },
    debug: (...args) => {
      // eslint-disable-next-line no-console, @typescript-eslint/no-unsafe-argument
      if (instance.enabled) console.debug(...args);
    },
    time: (label) => {
      // eslint-disable-next-line no-console
      if (instance.enabled) console.time(label);
    },
    timeEnd: (label) => {
      // eslint-disable-next-line no-console
      if (instance.enabled) console.timeEnd(label);
    },
    dir: (object, options) => {
      // eslint-disable-next-line no-console
      if (instance.enabled) console.dir(object, options);
    },
    table: (tabularData, properties) => {
      // eslint-disable-next-line no-console
      if (instance.enabled) console.table(tabularData, properties);
    },
    count: (label) => {
      // eslint-disable-next-line no-console
      if (instance.enabled) console.count(label);
    },
    countReset: (label) => {
      // eslint-disable-next-line no-console
      if (instance.enabled) console.countReset(label);
    },
    assert: (value, message, ...optionalParams) => {
      // eslint-disable-next-line no-console, @typescript-eslint/no-unsafe-argument
      if (instance.enabled) console.assert(value, message, ...optionalParams);
    },
    timeLog: (label, ...data) => {
      // eslint-disable-next-line no-console, @typescript-eslint/no-unsafe-argument
      if (instance.enabled) console.timeLog(label, ...data);
    },
    trace: (message, ...optionalParams) => {
      // eslint-disable-next-line no-console, @typescript-eslint/no-unsafe-argument
      if (instance.enabled) console.trace(message, ...optionalParams);
    },
  };

  return instance;
};
