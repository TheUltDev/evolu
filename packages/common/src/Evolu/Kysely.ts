import {
  CreateIndexBuilder,
  DummyDriver,
  Kysely,
  SqliteAdapter,
  SqliteQueryCompiler,
} from "kysely";
import { DbIndex } from "./DbSchema.js";

// https://kysely.dev/docs/recipes/splitting-query-building-and-execution
export const kysely = new Kysely({
  dialect: {
    createAdapter: () => new SqliteAdapter(),
    createDriver: () => new DummyDriver(),
    createIntrospector() {
      throw new Error("Not implemeneted");
    },
    createQueryCompiler: () => new SqliteQueryCompiler(),
  },
});

export type DbIndexesBuilder = (
  create: (indexName: string) => CreateIndexBuilder,
) => ReadonlyArray<CreateIndexBuilder<any>>;

const createIndex = kysely.schema.createIndex.bind(kysely.schema);

export const createIndexes = (
  indexes?: DbIndexesBuilder,
): ReadonlyArray<DbIndex> => {
  if (!indexes) return [];
  return indexes(createIndex).map(
    (index): DbIndex => ({
      name: index.toOperationNode().name.name,
      sql: index.compile().sql,
    }),
  );
};
