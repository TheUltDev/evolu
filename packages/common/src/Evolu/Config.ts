import { ConsoleConfig } from "../Console.js";
import { getOrThrow } from "../Result.js";
import { Mnemonic, SimpleName } from "../Type.js";
import type { DbIndexesBuilder } from "./Kysely.js";

export interface Config extends ConsoleConfig {
  /**
   * The name of Evolu instances. Evolu is multitenant - it can run more
   * instances concurrently. Every Evolu instance has to have its own unique
   * name. Database files are separated and invisible to each other.
   *
   * The default value is: `Evolu`.
   *
   * ### Example
   *
   * ```ts
   * // name: getOrThrow(SimpleName.from("MyApp"))
   * ```
   */
  readonly name: SimpleName;

  /**
   * URL for Evolu sync and backup server.
   *
   * The default value is `https://evolu.world`.
   */
  readonly syncUrl: string;

  /**
   * URL to reload browser tabs after reset or restore.
   *
   * The default value is `/`.
   */
  readonly reloadUrl: string;

  /**
   * Maximum physical clock drift allowed in ms.
   *
   * The default value is 5 * 60 * 1000 (5 minutes).
   */
  readonly maxDrift: number;

  /**
   * Use the `indexes` option to define SQLite indexes.
   *
   * Table and column names are not typed because Kysely doesn't support it.
   *
   * https://medium.com/@JasonWyatt/squeezing-performance-from-sqlite-indexes-indexes-c4e175f3c346
   *
   * ### Example
   *
   * ```ts
   * const evolu = createEvolu(evoluReactDeps)(Schema, {
   *   indexes: (create) => [
   *     create("todoCreatedAt").on("todo").column("createdAt"),
   *     create("todoCategoryCreatedAt")
   *       .on("todoCategory")
   *       .column("createdAt"),
   *   ],
   * });
   * ```
   */
  readonly indexes?: DbIndexesBuilder;

  /**
   * Use this option to create Evolu with the specified mnemonic. If omitted,
   * the mnemonic will be autogenerated. That should be the default behavior
   * until special UX requirements are needed (e.g., multitenancy).
   */
  readonly mnemonic?: Mnemonic;
}

export interface ConfigDep {
  readonly config: Config;
}

export const defaultConfig: Config = {
  name: getOrThrow(SimpleName.fromParent("Evolu")),
  syncUrl: "https://evolu.world",
  reloadUrl: "/",
  maxDrift: 5 * 60 * 1000,
  enableLogging: false,
};
