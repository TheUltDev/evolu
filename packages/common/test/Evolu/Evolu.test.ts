import { describe, expect, expectTypeOf, test, vi } from "vitest";
import { createConsole } from "../../src/Console.js";
import { createEvolu } from "../../src/Evolu/Evolu.js";
import { getOrThrow } from "../../src/Result.js";
import { SqliteBoolean } from "../../src/Sqlite.js";
import {
  Boolean,
  id,
  InferType,
  maxLength,
  NonEmptyString,
  nullOr,
  SimpleName,
} from "../../src/Type.js";
import {
  testCreateId,
  testNanoIdLib,
  testSimpleName,
  testTime,
} from "../_deps.js";
import {
  ValidateColumnTypes,
  ValidateIdColumnType,
  ValidateNoDefaultColumns,
  ValidateSchemaHasId,
} from "../../src/Evolu/Schema.js";

const TodoId = id("Todo");
type TodoId = InferType<typeof TodoId>;

const TodoCategoryId = id("TodoCategory");
type TodoCategoryId = InferType<typeof TodoCategoryId>;

const NonEmptyString50 = maxLength(50)(NonEmptyString);
type NonEmptyString50 = InferType<typeof NonEmptyString50>;

const Schema = {
  todo: {
    id: TodoId,
    title: NonEmptyString50,
    isCompleted: nullOr(SqliteBoolean),
    categoryId: nullOr(TodoCategoryId),
  },
  todoCategory: {
    id: TodoCategoryId,
    name: NonEmptyString50,
  },
};

const mockDeps = () => {
  const dbWorker = {
    onMessage: vi.fn(),
    postMessage: vi.fn(),
  };

  return {
    createDbWorker: () => dbWorker,
    time: testTime,
    nanoIdLib: testNanoIdLib,
    console: createConsole(),
    createAppState: () => ({
      reset: vi.fn(),
    }),
  };
};

let instancesCount = 0;

const setupEvoluTest = () => {
  const deps = mockDeps();
  const evolu = createEvolu(deps)(Schema, {
    name: getOrThrow(SimpleName.from(`instance${instancesCount++}`)),
  });
  const dbWorker = deps.createDbWorker();

  return { deps, evolu, dbWorker };
};

test("init postMessage call", () => {
  const { dbWorker } = setupEvoluTest();

  expect(dbWorker.postMessage.mock.calls).toMatchInlineSnapshot(`
    [
      [
        {
          "config": {
            "enableLogging": false,
            "maxDrift": 300000,
            "name": "instance0",
            "reloadUrl": "/",
            "syncUrl": "https://free.evoluhq.com",
          },
          "dbSchema": {
            "indexes": [],
            "tables": [
              {
                "columns": [
                  "title",
                  "isCompleted",
                  "categoryId",
                ],
                "name": "todo",
              },
              {
                "columns": [
                  "name",
                ],
                "name": "todoCategory",
              },
            ],
          },
          "initialData": [],
          "type": "init",
        },
      ],
    ]
  `);
});

test("insert should validate input and call postMessage", async () => {
  const { evolu, dbWorker } = setupEvoluTest();

  const validTodo = {
    title: "Test Todo",
  };

  const result = evolu.insert("todo", validTodo);

  expect(result.ok).toBe(true);
  expect(result.ok && result.value.id).toMatchInlineSnapshot(
    `"esTwHwplqLBSE8Ou8ffX4"`,
  );

  await new Promise((resolve) => setTimeout(resolve, 0));

  expect(dbWorker.postMessage.mock.calls[1]).toMatchInlineSnapshot(`
    [
      {
        "changes": [
          {
            "id": "esTwHwplqLBSE8Ou8ffX4",
            "table": "todo",
            "values": {
              "title": "Test Todo",
            },
          },
        ],
        "onCompleteIds": [],
        "subscribedQueries": [],
        "tabId": "s9rNnUeri8bzhS5AX_mRy",
        "type": "mutate",
      },
    ]
  `);
});

test("insert should reject invalid input", () => {
  const { evolu } = setupEvoluTest();

  // Empty title (violates NonEmptyString constraint)
  const invalidTodo = {
    title: "",
  };

  const result = evolu.insert("todo", invalidTodo);

  expect(result).toMatchInlineSnapshot(`
    {
      "error": {
        "reason": {
          "errors": {
            "title": {
              "min": 1,
              "type": "MinLength",
              "value": "",
            },
          },
          "kind": "Props",
        },
        "type": "Object",
        "value": {
          "title": "",
        },
      },
      "ok": false,
    }
  `);
});

test("update should validate input and call postMessage", async () => {
  const { evolu, dbWorker } = setupEvoluTest();

  const testId = testCreateId();

  // Valid update
  const validUpdate = {
    id: testId,
    title: "Updated Todo",
  };

  const result = evolu.update("todo", validUpdate);

  expect(result.ok).toBe(true);
  expect(result.ok && result.value.id).toBe(testId);

  await new Promise((resolve) => setTimeout(resolve, 0));

  expect(dbWorker.postMessage.mock.calls[1]).toMatchInlineSnapshot(`
    [
      {
        "changes": [
          {
            "id": "e37rn9q4lwirRDIx1Y9e4",
            "table": "todo",
            "values": {
              "title": "Updated Todo",
            },
          },
        ],
        "onCompleteIds": [],
        "subscribedQueries": [],
        "tabId": "s9rNnUeri8bzhS5AX_mRy",
        "type": "mutate",
      },
    ]
  `);
});

test("update should reject invalid input", () => {
  const { evolu } = setupEvoluTest();

  // Missing id
  const invalidUpdate = {
    title: "Updated Todo",
  };

  // @ts-expect-error - Testing runtime validation
  const result = evolu.update("todo", invalidUpdate);

  // Should return error
  expect(result).toMatchInlineSnapshot(`
    {
      "error": {
        "reason": {
          "errors": {
            "id": {
              "type": "String",
              "value": undefined,
            },
          },
          "kind": "Props",
        },
        "type": "Object",
        "value": {
          "title": "Updated Todo",
        },
      },
      "ok": false,
    }
  `);
});

test("upsert should validate input and call postMessage", async () => {
  const { evolu, dbWorker } = setupEvoluTest();

  const testId = testCreateId();

  // Valid upsert
  const validUpsert = {
    id: testId,
    title: "Upserted Todo",
  };

  const result = evolu.upsert("todo", validUpsert);

  expect(result.ok).toBe(true);
  expect(result.ok && result.value.id).toBe(testId);

  await new Promise((resolve) => setTimeout(resolve, 0));

  expect(dbWorker.postMessage.mock.calls[1]).toMatchInlineSnapshot(`
    [
      {
        "changes": [
          {
            "id": "ZUN-T4MZhx0p9fO9i2LT9",
            "table": "todo",
            "values": {
              "title": "Upserted Todo",
            },
          },
        ],
        "onCompleteIds": [],
        "subscribedQueries": [],
        "tabId": "s9rNnUeri8bzhS5AX_mRy",
        "type": "mutate",
      },
    ]
  `);
});

test("upsert should reject invalid input", () => {
  const { evolu } = setupEvoluTest();

  const testId = testCreateId();

  const invalidUpsert = {
    id: testId,
    title: "",
  };

  const result = evolu.upsert("todo", invalidUpsert);

  expect(result).toMatchInlineSnapshot(`
    {
      "error": {
        "reason": {
          "errors": {
            "title": {
              "min": 1,
              "type": "MinLength",
              "value": "",
            },
          },
          "kind": "Props",
        },
        "type": "Object",
        "value": {
          "id": "yKrUtl5RZ4mz12WZ_SQtm",
          "title": "",
        },
      },
      "ok": false,
    }
  `);
});

test("mutations should be processed in microtask queue", async () => {
  const { evolu, dbWorker } = setupEvoluTest();

  // Queue multiple mutations
  evolu.insert("todo", { title: "Todo 1" });
  evolu.insert("todo", { title: "Todo 2" });
  evolu.insert("todo", { title: "Todo 3" });

  await new Promise((resolve) => setTimeout(resolve, 0));

  // Only one postMessage call should happen with all changes
  expect(dbWorker.postMessage).toHaveBeenCalledTimes(2); // 1 for init, 1 for mutations
  expect(dbWorker.postMessage.mock.calls[1]).toMatchInlineSnapshot(`
    [
      {
        "changes": [
          {
            "id": "fIT5Ci6Vt_fajhVHzg5iu",
            "table": "todo",
            "values": {
              "title": "Todo 1",
            },
          },
          {
            "id": "oSfTGGJoCBJ9xgjl8Ky6B",
            "table": "todo",
            "values": {
              "title": "Todo 2",
            },
          },
          {
            "id": "pXftCAjUskTmno2dqqyjQ",
            "table": "todo",
            "values": {
              "title": "Todo 3",
            },
          },
        ],
        "onCompleteIds": [],
        "subscribedQueries": [],
        "tabId": "s9rNnUeri8bzhS5AX_mRy",
        "type": "mutate",
      },
    ]
  `);
});

test("mutation with onlyValidate should not call postMessage", async () => {
  const { evolu, dbWorker } = setupEvoluTest();

  evolu.insert("todo", { title: "Validation only" }, { onlyValidate: true });

  await new Promise((resolve) => setTimeout(resolve, 0));

  // Only init should be called, not the mutation
  expect(dbWorker.postMessage).toHaveBeenCalledTimes(1);
});

test("mutations should fail as a transaction when any mutation fails", async () => {
  const { evolu, dbWorker } = setupEvoluTest();

  // Queue valid and invalid mutations
  evolu.insert("todo", { title: "Valid Todo" });
  evolu.insert("todo", { title: "" }); // Invalid - empty title
  evolu.insert("todo", { title: "Another Valid Todo" });

  await new Promise((resolve) => setTimeout(resolve, 0));

  // Only init should be called, not the mutations since one failed
  expect(dbWorker.postMessage).toHaveBeenCalledTimes(1);
});

describe("EvoluSchema validation", () => {
  test("schema without id column", () => {
    const deps = mockDeps();

    const SchemaWithoutId = {
      todo: {
        // Missing id column - should cause TypeScript error
        title: NonEmptyString50,
      },
    };

    // Type-level assertion for the exact error message
    type ValidationResult = ValidateSchemaHasId<typeof SchemaWithoutId>;
    expectTypeOf<ValidationResult>().toEqualTypeOf<'❌ Schema Error: Table "todo" is missing required id column.'>();

    // @ts-expect-error - Schema validation should catch missing id column
    createEvolu(deps)(SchemaWithoutId, {
      name: testSimpleName,
    });
  });

  test("schema with default column createdAt", () => {
    const deps = mockDeps();

    const SchemaWithDefaultColumn = {
      todo: {
        id: TodoId,
        createdAt: NonEmptyString50,
      },
    };

    // Type-level assertion for the exact error message
    type ValidationResult = ValidateNoDefaultColumns<
      typeof SchemaWithDefaultColumn
    >;
    expectTypeOf<ValidationResult>().toEqualTypeOf<'❌ Schema Error: Table "todo" uses default column name "createdAt". Default columns (createdAt, updatedAt, isDeleted) are added automatically.'>();

    // @ts-expect-error - Schema validation should catch default column name
    createEvolu(deps)(SchemaWithDefaultColumn, {
      name: testSimpleName,
    });
  });

  test("schema with default column updatedAt", () => {
    const deps = mockDeps();

    const SchemaWithDefaultColumn = {
      todo: {
        id: TodoId,
        updatedAt: NonEmptyString50,
      },
    };

    // Type-level assertion for the exact error message
    type ValidationResult = ValidateNoDefaultColumns<
      typeof SchemaWithDefaultColumn
    >;
    expectTypeOf<ValidationResult>().toEqualTypeOf<'❌ Schema Error: Table "todo" uses default column name "updatedAt". Default columns (createdAt, updatedAt, isDeleted) are added automatically.'>();

    // @ts-expect-error - Schema validation should catch default column name
    createEvolu(deps)(SchemaWithDefaultColumn, {
      name: testSimpleName,
    });
  });

  test("schema with default column isDeleted", () => {
    const deps = mockDeps();

    const SchemaWithDefaultColumn = {
      todo: {
        id: TodoId,
        isDeleted: NonEmptyString50,
      },
    };

    // Type-level assertion for the exact error message
    type ValidationResult = ValidateNoDefaultColumns<
      typeof SchemaWithDefaultColumn
    >;
    expectTypeOf<ValidationResult>().toEqualTypeOf<'❌ Schema Error: Table "todo" uses default column name "isDeleted". Default columns (createdAt, updatedAt, isDeleted) are added automatically.'>();

    // @ts-expect-error - Schema validation should catch default column name
    createEvolu(deps)(SchemaWithDefaultColumn, {
      name: testSimpleName,
    });
  });

  test("schema with non-branded id column", () => {
    const deps = mockDeps();

    const SchemaWithInvalidId = {
      todo: {
        id: NonEmptyString50,
        title: NonEmptyString50,
      },
    };

    // Type-level assertion for the exact error message
    type ValidationResult = ValidateIdColumnType<typeof SchemaWithInvalidId>;
    expectTypeOf<ValidationResult>().toEqualTypeOf<'❌ Schema Error: Table "todo" id column must be a branded ID type (created with id("todo")).'>();

    // @ts-expect-error - Schema validation should catch non-branded id column
    createEvolu(deps)(SchemaWithInvalidId, {
      name: testSimpleName,
    });
  });

  test("schema with incompatible column type", () => {
    const deps = mockDeps();

    const SchemaWithInvalidType = {
      todo: {
        id: TodoId,
        title: NonEmptyString50,
        invalidColumn: Boolean, // Boolean is not compatible with SQLite
      },
    };

    // Type-level assertion for the exact error message
    type ValidationResult = ValidateColumnTypes<typeof SchemaWithInvalidType>;
    expectTypeOf<ValidationResult>().toEqualTypeOf<'❌ Schema Error: Table "todo" column "invalidColumn" type is not compatible with SQLite. Column types must extend SqliteValue (string, number, Uint8Array, or null).'>();

    // @ts-expect-error - Schema validation should catch incompatible column type
    createEvolu(deps)(SchemaWithInvalidType, {
      name: testSimpleName,
    });
  });
});
