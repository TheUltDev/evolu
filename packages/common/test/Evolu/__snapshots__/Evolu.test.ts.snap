// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`initialAppOwner should use provided owner 1`] = `
{
  "schema": {
    "indexes": [
      {
        "name": "evolu_history_ownerId_timestamp",
        "sql": "create index evolu_history_ownerId_timestamp on evolu_history (
          "ownerId",
          "timestamp"
        )",
      },
      {
        "name": "evolu_history_ownerId_table_id_column_timestampDesc",
        "sql": "create unique index evolu_history_ownerId_table_id_column_timestampDesc on evolu_history (
          "ownerId",
          "table",
          "id",
          "column",
          "timestamp" desc
        )",
      },
      {
        "name": "evolu_timestamp_index",
        "sql": "create index evolu_timestamp_index on evolu_timestamp (
        "ownerId",
        "l",
        "t",
        "h1",
        "h2",
        "c"
      )",
      },
    ],
    "tables": [
      {
        "columns": [
          "protocolVersion",
        ],
        "name": "evolu_version",
      },
      {
        "columns": [
          "clock",
          "appOwnerId",
          "appOwnerEncryptionKey",
          "appOwnerWriteKey",
          "appOwnerMnemonic",
        ],
        "name": "evolu_config",
      },
      {
        "columns": [
          "ownerId",
          "table",
          "id",
          "column",
          "timestamp",
          "value",
        ],
        "name": "evolu_history",
      },
      {
        "columns": [
          "id",
          "title",
          "isCompleted",
          "categoryId",
          "createdAt",
          "updatedAt",
          "isDeleted",
        ],
        "name": "todo",
      },
      {
        "columns": [
          "id",
          "name",
          "createdAt",
          "updatedAt",
          "isDeleted",
        ],
        "name": "todoCategory",
      },
      {
        "columns": [
          "ownerId",
          "t",
          "h1",
          "h2",
          "c",
          "l",
        ],
        "name": "evolu_timestamp",
      },
    ],
  },
  "tables": [
    {
      "name": "evolu_version",
      "rows": [
        {
          "protocolVersion": 0,
        },
      ],
    },
    {
      "name": "evolu_config",
      "rows": [
        {
          "appOwnerEncryptionKey": {
            "data": [
              214,
              142,
              243,
              254,
              221,
              13,
              246,
              19,
              136,
              198,
              193,
              14,
              183,
              237,
              206,
              51,
              171,
              193,
              65,
              231,
              26,
              97,
              31,
              142,
              157,
              180,
              17,
              113,
              106,
              171,
              129,
              26,
            ],
            "type": "Buffer",
          },
          "appOwnerId": "m4Swbg7PvGPpEyB_q4JY9",
          "appOwnerMnemonic": "call brass keen rough true spy dream robot useless ignore anxiety badge",
          "appOwnerWriteKey": {
            "data": [
              193,
              200,
              155,
              104,
              173,
              156,
              172,
              55,
              27,
              230,
              228,
              51,
              167,
              204,
              43,
              8,
            ],
            "type": "Buffer",
          },
          "clock": "1970-01-01T00:00:00.000Z-0000-2840b80279bdc6f6",
        },
      ],
    },
    {
      "name": "evolu_history",
      "rows": [],
    },
    {
      "name": "todo",
      "rows": [],
    },
    {
      "name": "todoCategory",
      "rows": [],
    },
    {
      "name": "evolu_timestamp",
      "rows": [],
    },
  ],
}
`;

exports[`onInit callback should be called with correct parameters and can seed initial data 1`] = `
{
  "schema": {
    "indexes": [
      {
        "name": "evolu_history_ownerId_timestamp",
        "sql": "create index evolu_history_ownerId_timestamp on evolu_history (
          "ownerId",
          "timestamp"
        )",
      },
      {
        "name": "evolu_history_ownerId_table_id_column_timestampDesc",
        "sql": "create unique index evolu_history_ownerId_table_id_column_timestampDesc on evolu_history (
          "ownerId",
          "table",
          "id",
          "column",
          "timestamp" desc
        )",
      },
      {
        "name": "evolu_timestamp_index",
        "sql": "create index evolu_timestamp_index on evolu_timestamp (
        "ownerId",
        "l",
        "t",
        "h1",
        "h2",
        "c"
      )",
      },
    ],
    "tables": [
      {
        "columns": [
          "protocolVersion",
        ],
        "name": "evolu_version",
      },
      {
        "columns": [
          "clock",
          "appOwnerId",
          "appOwnerEncryptionKey",
          "appOwnerWriteKey",
          "appOwnerMnemonic",
        ],
        "name": "evolu_config",
      },
      {
        "columns": [
          "ownerId",
          "table",
          "id",
          "column",
          "timestamp",
          "value",
        ],
        "name": "evolu_history",
      },
      {
        "columns": [
          "id",
          "title",
          "isCompleted",
          "categoryId",
          "createdAt",
          "updatedAt",
          "isDeleted",
        ],
        "name": "todo",
      },
      {
        "columns": [
          "id",
          "name",
          "createdAt",
          "updatedAt",
          "isDeleted",
        ],
        "name": "todoCategory",
      },
      {
        "columns": [
          "ownerId",
          "t",
          "h1",
          "h2",
          "c",
          "l",
        ],
        "name": "evolu_timestamp",
      },
    ],
  },
  "tables": [
    {
      "name": "evolu_version",
      "rows": [
        {
          "protocolVersion": 0,
        },
      ],
    },
    {
      "name": "evolu_config",
      "rows": [
        {
          "appOwnerEncryptionKey": {
            "data": [
              214,
              142,
              243,
              254,
              221,
              13,
              246,
              19,
              136,
              198,
              193,
              14,
              183,
              237,
              206,
              51,
              171,
              193,
              65,
              231,
              26,
              97,
              31,
              142,
              157,
              180,
              17,
              113,
              106,
              171,
              129,
              26,
            ],
            "type": "Buffer",
          },
          "appOwnerId": "m4Swbg7PvGPpEyB_q4JY9",
          "appOwnerMnemonic": "call brass keen rough true spy dream robot useless ignore anxiety badge",
          "appOwnerWriteKey": {
            "data": [
              193,
              200,
              155,
              104,
              173,
              156,
              172,
              55,
              27,
              230,
              228,
              51,
              167,
              204,
              43,
              8,
            ],
            "type": "Buffer",
          },
          "clock": "1970-01-01T00:00:00.011Z-0001-47ebc9cd4ddaff08",
        },
      ],
    },
    {
      "name": "evolu_history",
      "rows": [
        {
          "column": "name",
          "id": {
            "data": [
              102,
              222,
              231,
              157,
              156,
              119,
              15,
              80,
              10,
              111,
              59,
              188,
              195,
              208,
              227,
              128,
            ],
            "type": "Buffer",
          },
          "ownerId": {
            "data": [
              155,
              132,
              176,
              110,
              14,
              207,
              188,
              99,
              233,
              19,
              32,
              127,
              171,
              130,
              88,
              244,
            ],
            "type": "Buffer",
          },
          "table": "todoCategory",
          "timestamp": {
            "data": [
              0,
              0,
              0,
              0,
              0,
              11,
              0,
              0,
              71,
              235,
              201,
              205,
              77,
              218,
              255,
              8,
            ],
            "type": "Buffer",
          },
          "value": "Not Urgent",
        },
        {
          "column": "createdAt",
          "id": {
            "data": [
              102,
              222,
              231,
              157,
              156,
              119,
              15,
              80,
              10,
              111,
              59,
              188,
              195,
              208,
              227,
              128,
            ],
            "type": "Buffer",
          },
          "ownerId": {
            "data": [
              155,
              132,
              176,
              110,
              14,
              207,
              188,
              99,
              233,
              19,
              32,
              127,
              171,
              130,
              88,
              244,
            ],
            "type": "Buffer",
          },
          "table": "todoCategory",
          "timestamp": {
            "data": [
              0,
              0,
              0,
              0,
              0,
              11,
              0,
              0,
              71,
              235,
              201,
              205,
              77,
              218,
              255,
              8,
            ],
            "type": "Buffer",
          },
          "value": "1970-01-01T00:00:00.009Z",
        },
        {
          "column": "title",
          "id": {
            "data": [
              186,
              111,
              116,
              220,
              155,
              7,
              109,
              90,
              172,
              0,
              91,
              44,
              202,
              207,
              217,
              144,
            ],
            "type": "Buffer",
          },
          "ownerId": {
            "data": [
              155,
              132,
              176,
              110,
              14,
              207,
              188,
              99,
              233,
              19,
              32,
              127,
              171,
              130,
              88,
              244,
            ],
            "type": "Buffer",
          },
          "table": "todo",
          "timestamp": {
            "data": [
              0,
              0,
              0,
              0,
              0,
              11,
              0,
              1,
              71,
              235,
              201,
              205,
              77,
              218,
              255,
              8,
            ],
            "type": "Buffer",
          },
          "value": "Try React Suspense",
        },
        {
          "column": "categoryId",
          "id": {
            "data": [
              186,
              111,
              116,
              220,
              155,
              7,
              109,
              90,
              172,
              0,
              91,
              44,
              202,
              207,
              217,
              144,
            ],
            "type": "Buffer",
          },
          "ownerId": {
            "data": [
              155,
              132,
              176,
              110,
              14,
              207,
              188,
              99,
              233,
              19,
              32,
              127,
              171,
              130,
              88,
              244,
            ],
            "type": "Buffer",
          },
          "table": "todo",
          "timestamp": {
            "data": [
              0,
              0,
              0,
              0,
              0,
              11,
              0,
              1,
              71,
              235,
              201,
              205,
              77,
              218,
              255,
              8,
            ],
            "type": "Buffer",
          },
          "value": "Zt7nnZx3D1AKbzu8w9Djg",
        },
        {
          "column": "createdAt",
          "id": {
            "data": [
              186,
              111,
              116,
              220,
              155,
              7,
              109,
              90,
              172,
              0,
              91,
              44,
              202,
              207,
              217,
              144,
            ],
            "type": "Buffer",
          },
          "ownerId": {
            "data": [
              155,
              132,
              176,
              110,
              14,
              207,
              188,
              99,
              233,
              19,
              32,
              127,
              171,
              130,
              88,
              244,
            ],
            "type": "Buffer",
          },
          "table": "todo",
          "timestamp": {
            "data": [
              0,
              0,
              0,
              0,
              0,
              11,
              0,
              1,
              71,
              235,
              201,
              205,
              77,
              218,
              255,
              8,
            ],
            "type": "Buffer",
          },
          "value": "1970-01-01T00:00:00.009Z",
        },
      ],
    },
    {
      "name": "todo",
      "rows": [
        {
          "categoryId": "Zt7nnZx3D1AKbzu8w9Djg",
          "createdAt": "1970-01-01T00:00:00.009Z",
          "id": "um903JsHbVqsAFssys_Zk",
          "isCompleted": null,
          "isDeleted": null,
          "title": "Try React Suspense",
          "updatedAt": "1970-01-01T00:00:00.011Z",
        },
      ],
    },
    {
      "name": "todoCategory",
      "rows": [
        {
          "createdAt": "1970-01-01T00:00:00.009Z",
          "id": "Zt7nnZx3D1AKbzu8w9Djg",
          "isDeleted": null,
          "name": "Not Urgent",
          "updatedAt": "1970-01-01T00:00:00.011Z",
        },
      ],
    },
    {
      "name": "evolu_timestamp",
      "rows": [
        {
          "c": 1,
          "h1": 259761259403146,
          "h2": 141925445732367,
          "l": 2,
          "ownerId": {
            "data": [
              155,
              132,
              176,
              110,
              14,
              207,
              188,
              99,
              233,
              19,
              32,
              127,
              171,
              130,
              88,
              244,
            ],
            "type": "Buffer",
          },
          "t": {
            "data": [
              0,
              0,
              0,
              0,
              0,
              11,
              0,
              0,
              71,
              235,
              201,
              205,
              77,
              218,
              255,
              8,
            ],
            "type": "Buffer",
          },
        },
        {
          "c": 1,
          "h1": 167102060756421,
          "h2": 202046611856140,
          "l": 1,
          "ownerId": {
            "data": [
              155,
              132,
              176,
              110,
              14,
              207,
              188,
              99,
              233,
              19,
              32,
              127,
              171,
              130,
              88,
              244,
            ],
            "type": "Buffer",
          },
          "t": {
            "data": [
              0,
              0,
              0,
              0,
              0,
              11,
              0,
              1,
              71,
              235,
              201,
              205,
              77,
              218,
              255,
              8,
            ],
            "type": "Buffer",
          },
        },
      ],
    },
  ],
}
`;
